test/language/statements/for-of/head-await-using-bound-names-fordecl-tdz.js

```diff
@@ -0,0 +1,17 @@
+// Copyright (C) 2023 Ron Buckton. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+es6id: 13.6.4.12_S2
+description: >
+    ForIn/Of: Bound names of ForDeclaration are in TDZ (for-of)
+flags: [async]
+includes: [asyncHelpers.js]
+features: [explicit-resource-management]
+---*/
+
+asyncTest(async function () {
+  await assert.throwsAsync(ReferenceError, async function() {
+    let x = { async [Symbol.asyncDispose]() { } };
+    for (await using x of [x]) {}
+  });
```

Comment:
I'm probably misunderstanding this due to my unfamiliarity with this part of the spec, but where is this ReferenceError thrown? I'm seeing in step 9.h.iii of ForIn/OfBodyEvaluation, we create a new declarative environment that already inherits the outer environment's `let x` binding in it from line 14. Then in ForDeclarationBindingInstantiation we do CreateImmutableBinding with `x` on the environment, which asserts that the environment does not already contain `x`.

Reply:
Due to TDZ, the `x` is looked up in the same scope as the `await using x`. Since `x` has been bound but not initialized, this results in a ReferenceError.

Binding occurs in [14.7.5.4 Runtime Semantics: ForDeclarationBindingInstantiation](https://arai-a.github.io/ecma262-compare/?pr=3000&id=sec-runtime-semantics-fordeclarationbindinginstantiation). The `ReferenceError` is thrown in [9.1.1.1.6 GetBindingValue (N, S)](https://tc39.es/ecma262/#sec-declarative-environment-records-getbindingvalue-n-s), step 2.